Paper: Variational Autoencoders For Semi-Supervised Deep Metric Learning

nssafir opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@stargaser<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@jecamil<!--end-reviewers-list-->

nssafir committed:
Add paper directory.

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

deniederhut commented on papers/safir_zain_22/main.rst:
```diff
@@ -0,0 +1,628 @@
+58,850, respectively) and in that they are both greyscale.
+
+.. figure:: figs/cropped_datasets.png
+   :scale: 45%
```
I'm not sure that docutils knows how to handle this line? Here's what I'm seeing from the server logs:

```
raise SystemMessage(msg, level)\ndocutils.utils.SystemMessage: <string>:457: (ERROR/3) Error in \"figure\" directive:\ninvalid option block.\n\n.. figure:: figs/cropped_datasets.png\n   :scale: 45%\n   :figclass: w\n   :align: center\n   Sample images from the MNIST (left) and OrganAMNIST of MedMNIST (right) datasets\n\nError building paper /tmp/tmp430tsinc/papers/safir_zain_22. Aborting.\n[X]
```

mepa commented on papers/safir_zain_22/main.rst:
```diff
@@ -0,0 +1,628 @@
+   :scale: 45%
+   :figclass: w
+   :align: center
+   Sample images from the MNIST (left) and OrganAMNIST of MedMNIST (right) datasets
```
```suggestion

Sample images from the MNIST (left) and OrganAMNIST of MedMNIST (right) datasets
```

The formatting of this line seems to be the source of the error that @deniederhut posted above.

mepa commented on papers/safir_zain_22/main.rst:
```diff
@@ -0,0 +1,628 @@
+would simply be to keep alternating between losses but at the level of each batch instead of each epoch.
+Another alternative, specifically for the MVAE, may be first training DML on labelled data, training a
+GMM on it’s outputs, and then using the GMM as the prior distribution for the VAE. Grosnit et al.
+(:citeyear:`grosnit2021high`) has defined a more complex training routines to balance the DML and unsupervised loss. If this
```
```suggestion
(:cite:`grosnit2021high`) has defined a more complex training routines to balance the DML and unsupervised loss. If this
```

mepa commented on papers/safir_zain_22/main.rst:
```diff
@@ -0,0 +1,628 @@
+amounts (roughly less than 100 samples) of labelled data and a sufficient quantity of unlabelled data.
+But an important caveat is that it is not convincing that the DML Autoencoder effectively combined
+the unsupervised and supervised losses to create a superior model, as a plain autoencoder (i.e. the DML
+Autoencoder with α = 1) outperforms the DML for the partial labels percentage of or less than 0.1% and
```
```suggestion
Autoencoder with :math:`\alpha = 1`) outperforms the DML for the partial labels percentage of or less than 0.1% and
```

mepa commented on papers/safir_zain_22/main.rst:
```diff
@@ -0,0 +1,628 @@
+   :align: center
+   
+   Table 4: Experiments performed on MVAE architecture across fully labelled MNIST dataset that trains
+   on objective function L = LU + γ ∗LS on fully supervised dataset. The best results for the classification
```
maybe this?
```suggestion
   on objective function :math:`L = LU + \gamma * LS` on fully supervised dataset. The best results for the classification
```

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"nssafir-2022"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmp5w8ie4e0'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\n<string>:457: (ERROR/3) Error in \"figure\" directive:\ninvalid option block.\n\n.. figure:: figs/cropped_datasets.png\n   :scale: 45%\n   :figclass: w\n   :align: center\n   Sample images from the MNIST (left) and OrganAMNIST of MedMNIST (right) datasets\n\nBuilding: safir_zain_22\nTraceback (most recent call last):\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2119, in run_directive\n    directive, option_presets))\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2171, in parse_directive_block\n    option_presets, option_spec, arg_block)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2206, in parse_directive_options\n    raise MarkupError(data)\ndocutils.parsers.rst.states.MarkupError: invalid option block\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"publisher/build_paper.py\", line 237, in <module>\n    build_paper(paper_id)\n  File \"publisher/build_paper.py\", line 222, in build_paper\n    rst2tex(in_path, out_path)\n  File \"publisher/build_paper.py\", line 92, in rst2tex\n    settings_overrides=settings)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/core.py\", line 416, in publish_string\n    enable_exit_status=enable_exit_status)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/core.py\", line 664, in publish_programmatically\n    output = pub.publish(enable_exit_status=enable_exit_status)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/core.py\", line 217, in publish\n    self.settings)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/readers/__init__.py\", line 72, in read\n    self.parse()\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/readers/__init__.py\", line 78, in parse\n    self.parser.parse(self.input, document)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/__init__.py\", line 191, in parse\n    self.statemachine.run(inputlines, document, inliner=self.inliner)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 171, in run\n    input_source=document['source'])\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 239, in run\n    context, state, transitions)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 460, in check_line\n    return method(match, context, next_state)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2988, in text\n    self.section(title.lstrip(), source, style, lineno + 1, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 327, in section\n    self.new_subsection(title, lineno, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 395, in new_subsection\n    node=section_node, match_titles=True)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 282, in nested_parse\n    node=node, match_titles=match_titles)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 196, in run\n    results = StateMachineWS.run(self, input_lines, input_offset)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 239, in run\n    context, state, transitions)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 460, in check_line\n    return method(match, context, next_state)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2753, in underline\n    self.section(title, source, style, lineno - 1, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 327, in section\n    self.new_subsection(title, lineno, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 395, in new_subsection\n    node=section_node, match_titles=True)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 282, in nested_parse\n    node=node, match_titles=match_titles)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 196, in run\n    results = StateMachineWS.run(self, input_lines, input_offset)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 239, in run\n    context, state, transitions)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 460, in check_line\n    return method(match, context, next_state)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2753, in underline\n    self.section(title, source, style, lineno - 1, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 327, in section\n    self.new_subsection(title, lineno, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 395, in new_subsection\n    node=section_node, match_titles=True)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 282, in nested_parse\n    node=node, match_titles=match_titles)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 196, in run\n    results = StateMachineWS.run(self, input_lines, input_offset)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 239, in run\n    context, state, transitions)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 460, in check_line\n    return method(match, context, next_state)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2753, in underline\n    self.section(title, source, style, lineno - 1, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 327, in section\n    self.new_subsection(title, lineno, messages)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 395, in new_subsection\n    node=section_node, match_titles=True)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 282, in nested_parse\n    node=node, match_titles=match_titles)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 196, in run\n    results = StateMachineWS.run(self, input_lines, input_offset)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 239, in run\n    context, state, transitions)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/statemachine.py\", line 460, in check_line\n    return method(match, context, next_state)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2326, in explicit_markup\n    nodelist, blank_finish = self.explicit_construct(match)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2338, in explicit_construct\n    return method(self, expmatch)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2081, in directive\n    directive_class, match, type_name, option_presets)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/parsers/rst/states.py\", line 2124, in run_directive\n    nodes.literal_block(block_text, block_text), line=lineno)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/utils/__init__.py\", line 228, in error\n    return self.system_message(self.ERROR_LEVEL, *args, **kwargs)\n  File \"/procbuild_env/lib/python3.6/site-packages/docutils/utils/__init__.py\", line 194, in system_message\n    raise SystemMessage(msg, level)\ndocutils.utils.SystemMessage: <string>:457: (ERROR/3) Error in \"figure\" directive:\ninvalid option block.\n\n.. figure:: figs/cropped_datasets.png\n   :scale: 45%\n   :figclass: w\n   :align: center\n   Sample images from the MNIST (left) and OrganAMNIST of MedMNIST (right) datasets\n\nError building paper /tmp/tmp5w8ie4e0/papers/safir_zain_22. Aborting.\n[X] Error code 1\nA build error occurred during the self.run_make_paper_script step.", "build_pdf_path"=>"", "build_status"=>"Build failed, make_paper.sh did not succeed", "build_timestamp"=>"27/05 03:21"}, "status"=>"fail"}}

deniederhut commented:
@scoobies mark server fail

deniederhut commented:
@scoobies assign @stargaser as editor

scoobies commented:
Assigned! @stargaser is now the editor

deniederhut commented:
Hey @nssafir ! Have you been able to get this paper building?

nssafir commented:
Thanks so much for the help and sorry for the late reply. I am currently
abroad and have spotty internet access, but should be able to try the new
build within the next 48 hours.

On Wed, Jun 1, 2022, 6:07 AM Dillon Niederhut ***@***.***>
wrote:

> Hey @nssafir <https://github.com/nssafir> ! Have you been able to get
> this paper building?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/695#issuecomment-1143068837>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALJMRYBYIT27JMRZAMSHJZ3VM3HXXANCNFSM5XBTRNMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

deniederhut commented:
@scoobies add @jecamil to reviewers

scoobies commented:
@jecamil added to the reviewers list!

nssafir committed:
Apply formatting suggestions from PR

Should fix paper build

Co-authored-by: Meghann Agarwal <agarwal.meghann@gmail.com>

nssafir commented:
@deniederhut looks like the changes worked, the paper is building now. Thanks for spotting that!

stargaser commented:
@scoobies mark server success

stargaser commented:
@scoobies mark paper needs review

stargaser commented:
@scoobies mark needs review

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

stargaser commented:
Hi @jecamil, have you had a chance to review this paper? The deadline is coming up tomorrow.

jecamil commented:
Hello @nssafir, 

First of all, thanks for making contributions to this very relevant topic! It is true that many sophisticated methods exist to deal with labelled data, but it frequently happens that we have small amount of labelled and huge amount of unlabelled data, so if we find better ways to include info from unlabelled part, it will be more than beneficial for the community.

My main comments about the paper are:

1. State of the art is not very clear, we can see that there is related work in this domain, but it is not clear which outcomes were achieved and what were the drawbacks of these works. This makes understanding of current status of the field and position of this paper within it hard. My suggestion is to be more clear with the other approaches proposed, what worked well in what they proposed, which performance increase was reported, what did not work? Are these approaches used and cited? What is still missing and how is this paper different than what is already there? 

2. We see that for each of the improvement trials there is no one conclusive outcome. However, it remains unclear what to learn from this lesson and how to interpret these results. Can you explain in more detail why it didn't work out in your opinion, which take-aways we can take from your work?

3. Comments related to the reported hyper-parameters: 

- In Algorithms 2&3- s is not introduced, what does it mean?

- Learning rate - same as for other hyper-parameters, it would be useful to list which other learning rates have been tried for better clarity of the paper.

- Partial labels percentage: values experimented with are listed to be: 0.01, 0.1, 10 and 100, however in Tables 1&2&3 results for pl of 100 are not present. It would be interesting to see them too for better understanding of the outcomes and take-aways. 

4. Wording around 'claims'- I propose to find a better word for claim and claims in the paper. The main reason is that claim seems too strong, especially given that this is essentially just a hypothesis that does not really hold given the results after. Some replacement suggestions are: assumptions, improvement proposals, etc. Example sentence: "Third, we claim that performing DML on the latent space of the VAE so that the DML task can be jointly optimized with the VAE to incorporate unlabelled data may help produce better representations" - better way to formulate this is to say our assumption is/we aim to investigate if, etc.

5. Enclosed algorithms are small and tricky to read, they should be enlarged for increased visibility. Also, enclosed sample images of Mnist and OrganMnist datasets are too small.

6. Some typos I noticed: visualilzations, hte, datset, hyperapameter, due to the the, at least for a the classfication task, will cluste raround,  repre- sentations, the as the. Also, there is no closed bracket in algorithm 5.

7. "We will evaluate the results by running each model on a test partition of data." - It is weird to use future tense in this context, better keep present tense.

Optional suggestions:

1. My general preference for abstracts is to also include a short info about the main takeaways in at least one sentence to make the overview more complete. 

2. Conclusion part is a mix between conclusions, next steps and proposal for improvements; one option is to name the section accordingly, i.e. conclusions and next steps.

3. Following paper could also be interesting for the Related work section: https://link.springer.com/chapter/10.1007/978-3-030-18590-9_50

I hope comments are clear, feel free to reach out if further clarification is needed!

stargaser commented:
@scoobies mark pending comment

nssafir committed:
Fix typos from feedback

"visualilzations, hte, datset, hyperapameter, due to the the, at least for a the classfication task, will cluste raround, repre- sentations, the as the"

nssafir committed:
Seperate conclusion into subsections

Added Conclusion and Future Work subsections to Conclusion section

nssafir committed:
Explain 's' hyperparam

nssafir committed:
Remove 100% as pl option

Never actually experimented with 100% partial labels due to error in software, should not have been listed.

nssafir committed:
Make one sentence present tense

nssafir committed:
Upload corrected algorithm 5

nssafir committed:
Elaborate slightly on learning rates

nssafir committed:
Replace "claim" terminology

Generally use term "improvement proposal" instead of term "claim"

stargaser commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper nssafir-2022 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

stargaser commented:
@scoobies build status

scoobies commented:
{"nssafir-2022"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpwdovxncl'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\n<string>:256: (WARNING/2) Title underline too short.\n\nImprovement Proposal 1\n+++++++++++++\n<string>:290: (WARNING/2) Title underline too short.\n\nImprovement Proposal 2\n+++++++++++++\n<string>:290: (WARNING/2) Title underline too short.\n\nImprovement Proposal 2\n+++++++++++++\n<string>:352: (WARNING/2) Title underline too short.\n\nImprovement Proposal 3\n+++++++++++++\n<string>:352: (WARNING/2) Title underline too short.\n\nImprovement Proposal 3\n+++++++++++++\n<string>:493: (WARNING/2) Title underline too short.\n\nImprovement Proposal 1 Results: Benefits of Reconstruction Loss\n+++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n<string>:493: (WARNING/2) Title underline too short.\n\nImprovement Proposal 1 Results: Benefits of Reconstruction Loss\n+++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n<string>:531: (WARNING/2) Title underline too short.\n\nImprovement Proposal 2 Results: Incorporating Inductive Bias with a Prior\n++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n<string>:531: (WARNING/2) Title underline too short.\n\nImprovement Proposal 2 Results: Incorporating Inductive Bias with a Prior\n++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n<string>:576: (WARNING/2) Title underline too short.\n\nImprovement Proposal 3 Results: Jointly Optimizing DML with VAE\n++++++++++++++++++++++++++++++++++++++++++++++++\n<string>:576: (WARNING/2) Title underline too short.\n\nImprovement Proposal 3 Results: Jointly Optimizing DML with VAE\n++++++++++++++++++++++++++++++++++++++++++++++++\nBuilding: safir_zain_22\n*** Warning: /tmp/tmpwdovxncl/publisher/../output/safir_zain_22/paper_stats.json does not exist.\nPDFLaTeX error output:\n================================================================================\nThis is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) (preloaded format=pdflatex)\n restricted \\write18 enabled.\nentering extended mode\n(./paper.tex\nLaTeX2e <2017-04-15>\nBabel <3.18> and hyphenation patterns for 3 language(s) loaded.\n(./IEEEtran.cls\nDocument Class: IEEEtran 2015/08/26 V1.8b by Michael Shell\n-- See the \"IEEEtran_HOWTO\" manual for usage information.\n-- http://www.michaelshell.org/tex/ieeetran/\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ot1ppl.fd)\n-- Using IEEE Computer Society mode.\n-- Using 8.5in x 11in (letter) paper.\n-- Using PDF output.\n-- This is a 10 point document.\n) (/usr/share/texlive/texmf-dist/tex/latex/cmap/cmap.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/base/ifthen.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/base/fontenc.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/t1enc.def)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1ppl.fd)<<t1.cmap>>)\n(/usr/share/texlive/texmf-dist/tex/latex/base/inputenc.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/utf8.def\n(/usr/share/texlive/texmf-dist/tex/latex/base/t1enc.dfu)\n(/usr/share/texlive/texmf-dist/tex/latex/base/ot1enc.dfu)\n(/usr/share/texlive/texmf-dist/tex/latex/base/omsenc.dfu)))\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsmath.sty\nFor additional information on amsmath, use the `?' option.\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amstext.sty\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsgen.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsbsy.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsopn.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/color.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics-cfg/color.cfg)\n(/usr/share/texlive/texmf-dist/tex/latex/graphics-def/pdftex.def))\n(/usr/share/texlive/texmf-dist/tex/latex/caption/caption.sty\n(/usr/share/texlive/texmf-dist/tex/latex/caption/caption3.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/keyval.sty))\n\nPackage caption Warning: Unsupported document class (or package) detected,\n(caption)                usage of the caption package is not recommended.\nSee the caption package documentation for explanation.\n\n) (/usr/share/texlive/texmf-dist/tex/latex/float/float.sty) (./scipy.sty\n(./status.sty (./draftwatermark.sty (./everypage.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/graphicx.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/graphics.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/trig.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)))))\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/mathptmx.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/helvet.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/courier.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/amsfonts.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/amssymb.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/tools/bm.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/hyperref.sty\n(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/hobsub-hyperref.sty\n(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/hobsub-generic.sty))\n(/usr/share/texlive/texmf-dist/tex/generic/ifxetex/ifxetex.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/auxhook.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/kvoptions.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/pd1enc.def)\n(/usr/share/texlive/texmf-dist/tex/latex/latexconfig/hyperref.cfg)\n(/usr/share/texlive/texmf-dist/tex/latex/url/url.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/hpdftex.def\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/rerunfilecheck.sty))\n\nPackage hyperref Warning: Option `pdftex' has already been used,\n(hyperref)                setting the option has no effect on input line 43.\n\n(/usr/share/texlive/texmf-dist/tex/latex/tools/longtable.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/multirow/multirow.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/booktabs/booktabs.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/base/textcomp.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/ts1enc.def\n(/usr/share/texlive/texmf-dist/tex/latex/base/ts1enc.dfu)))\n(/usr/share/texlive/texmf-dist/tex/latex/upquote/upquote.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/fancyvrb/fancyvrb.sty\nStyle option: `fancyvrb' v2.7a, with DG/SPQR fixes, and firstline=lastline fix \n<2008/02/07> (tvz)))\nNo file paper.aux.\n(/usr/share/texlive/texmf-dist/tex/latex/base/ts1cmr.fd)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1ptm.fd)\n-- Lines per column: 61 (exact).\n(/usr/share/texlive/texmf-dist/tex/context/base/mkii/supp-pdf.mkii\n[Loading MPS to PDF converter (version 2006.09.02).]\n) (/usr/share/texlive/texmf-dist/tex/latex/oberdiek/epstopdf-base.sty\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/grfext.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/latexconfig/epstopdf-sys.cfg))\n(/usr/share/texlive/texmf-dist/tex/latex/caption/ltcaption.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/nameref.sty\n(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/gettitlestring.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1phv.fd)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ot1ztmcm.fd)<<ot1.cmap>>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/omlztmcm.fd)<<oml.cmap>>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/omsztmcm.fd)<<oms.cmap>>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/omxztmcm.fd)<<omx.cmap>>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ot1ptm.fd)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/upzd.fd)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ts1ptm.fd) (./page_numbers.tex)\n\nLaTeX Warning: Citation `bilenko2004integrating' on page 1 undefined on input l\nine 257.\n\n\nLaTeX Warning: Citation `baghshah2009semi' on page 1 undefined on input line 26\n2.\n\n\nLaTeX Warning: Citation `wang2013semi' on page 1 undefined on input line 265.\n\n[1{/var/lib/texmf/fonts/map/pdftex/updmap/pdftex.map}]\n\nLaTeX Warning: Citation `dutta2021semi' on page 2 undefined on input line 275.\n\n\nLaTeX Warning: Citation `wu2020metric' on page 2 undefined on input line 279.\n\n\nLaTeX Warning: Citation `lin2018deep' on page 2 undefined on input line 293.\n\n\nLaTeX Warning: Citation `kulkarni2020deep' on page 2 undefined on input line 30\n2.\n\n\nLaTeX Warning: Citation `sohn2015learning' on page 2 undefined on input line 30\n8.\n\n\nLaTeX Warning: Citation `dahmani2019conditional' on page 2 undefined on input l\nine 309.\n\n\nLaTeX Warning: Citation `sohn2016improved' on page 2 undefined on input line 31\n1.\n\n\nLaTeX Warning: Citation `grosnit2021high' on page 2 undefined on input line 319\n.\n\n\nLaTeX Warning: Citation `schroff2015facenet' on page 2 undefined on input line \n363.\n\n\nLaTeX Warning: No positions in optional float specifier.\n               Default added (so using `tbp') on input line 376.\n\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1pcr.fd)\n! Argument of \\@xverbatim has an extra }.\n<inserted text> \n                \\par \nl.406 backrefs: }\n                 \n!  ==> Fatal error occurred, no output PDF file produced!\nTranscript written on paper.log.\n\n================================================================================\n[X] Paper build failed.\nA build error occurred during the self.retrieve_pdf step.", "build_pdf_path"=>"", "build_status"=>"Build failed, no pdf can be found", "build_timestamp"=>"25/06 17:49"}, "status"=>"fail"}}

nssafir committed:
Fix subtitle underlines too short warnings

nssafir committed:
Add short conclusion statement to abstract.

nssafir committed:
Attempt increasing size of algorithm 1

Doubled scale

nssafir committed:
Double size of pseudocode and dataset samples

nssafir committed:
Fix caption for cropped datasets

nssafir committed:
Discuss what did/didn't work in conclusion

nssafir committed:
Add SSCDML model source

nssafir committed:
Refer to Li 2019 paper in related work

nssafir commented:
@stargaser would you happen to know why the citations aren't rendering in the PDF? I've tried checking against some other papers that do have their citations rendering and it seemed like their RST for their citations is the same.

deniederhut commented:
> `:bibliography: ref`

@nssafir it looks like the file is named refs, plural?

nssafir committed:
Refer to correct bib filename

nssafir commented:
> > `:bibliography: ref`
> 
> @nssafir it looks like the file is named refs, plural?

that was it, thanks!

deniederhut commented:
@nssafir is this paper ready for re-review?

stargaser commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.3828935 is OK

MISSING DOIs

- 10.21437/interspeech.2019-2848 may be a valid DOI for title: Conditional variational auto-encoder for text-driven expressive audiovisual speech synthesis
- 10.1109/cvpr.2015.7298682 may be a valid DOI for title: Facenet: A unified embedding for face recognition and clustering
- 10.1109/cvpr.2005.202 may be a valid DOI for title: Learning a similarity metric discriminatively, with application to face verification
- 10.1109/cvpr.2017.713 may be a valid DOI for title: Sphereface: Deep hypersphere embedding for face recognition
- 10.1109/cvpr.2018.00552 may be a valid DOI for title: Cosface: Large margin cosine loss for deep face recognition
- 10.1109/cvpr.2019.00482 may be a valid DOI for title: Arcface: Additive angular margin loss for deep face recognition
- 10.1109/cvpr.2019.01108 may be a valid DOI for title: Adacos: Adaptively scaling cosine logits for effectively learning deep face representations
- 10.1142/9789811223334_0120 may be a valid DOI for title: Metric learning by similarity network for deep semi-supervised learning
- 10.1145/1015330.1015360 may be a valid DOI for title: Integrating constraints and metric learning in semi-supervised clustering
- 10.1016/j.patcog.2013.02.015 may be a valid DOI for title: Semi-supervised metric learning via topology preserving multiple semi-supervised assumptions
- 10.1007/978-3-030-93409-5_6 may be a valid DOI for title: Finding Cross-Border Collaborative Centres in Biopharma Patent Networks: A Clustering Comparison Approach Based on Adjusted Mutual Information
- 10.1371/journal.pone.0159161 may be a valid DOI for title: Analysis of network clustering algorithms and cluster quality metrics at scale
- 10.25080/majora-342d178e-017 may be a valid DOI for title: Towards an Unsupervised Spatiotemporal Representation of Cilia Video Using A Modular Generative Pipeline

INVALID DOIs

- https://doi.org/10.1016/j.ins.2021.05.016 is INVALID because of 'https://doi.org/' prefix
```

stargaser commented:
@nssafir would you please check over the references for missing DOIs, per the previous comment?

stargaser commented:
@nssafir another issue, the paper is slightly over the 8-page limit (not including references).

nssafir committed:
Add DOI information suggested in build logs

nssafir committed:
Remove hardware section and scale down images

Reduce size of pseudocode figures

nssafir committed:
Cut words in future work

nssafir committed:
Remove typos in abstract

nssafir committed:
Include benchmark information from Li paper

nssafir commented:
@stargaser thank you for bringing those 2 issues to my attention, I think I was able to resolve both of them. @deniederhut sorry I had thought I replied to your comment a couple days ago, but the paper should be ready for re-review now.

deniederhut commented:
Awesome! @jecamil do you feel this paper is now ready for inclusion in the proceedings?

jecamil commented:
@nssafir Thanks for doing the changes as we discussed. I see that the point that is not fully covered so far is related to the state of the art situation. If you could be more clear with the other approaches proposed, what worked well in what they proposed, which performance increase was reported, what did not work, and how much are these approaches used/cited. Any information you can add to this part would be useful to better position contribution of this paper.

nssafir committed:
Add citations used in SOTA discussion

nssafir committed:
Add more discussion on SOTA

nssafir commented:
Thank you for your feedback @jecamil , I just added a paragraph in the last commit (last paragraph of the S"emi-Supervised Metric Learning" section) that should address the questions you brought up, let me know what you think and if I can change anything.

jecamil commented:
@nssafir Thanks for the quick response. Now it all looks good to me, you have my green light to move on! Looking forward seeing the paper out :)

deniederhut commented:
@stargaser is this paper ready for publication?

stargaser commented:
@scoobies mark ready

deniederhut commented on papers/safir_zain_22/refs.bib:
```diff
@@ -0,0 +1,401 @@
+  year={2018},
+  publisher={IEEE}
+}
+@article{ren2019semi,
```
Is this missing a DOI?

deniederhut commented on papers/safir_zain_22/refs.bib:
```diff
@@ -0,0 +1,401 @@
+  year={2019},
+  organization={Springer}
+}
+@article{arshad2018semi,
```
Is this also missing a DOI?

deniederhut commented on papers/safir_zain_22/refs.bib:
```diff
@@ -0,0 +1,401 @@
+url={https://www.wandb.com/},
+author = {Biewald, Lukas},
+}
+@inproceedings{li2019semi,
```
Is this also also also missing a DOI?

deniederhut replied:
Thanks for adding the DOI! It doesn't appear to be showing up in the rendered pdf though?

<img width="367" alt="Screen Shot 2022-07-08 at 6 58 15 PM" src="https://user-images.githubusercontent.com/6868396/178083225-1f2fd012-e115-41bf-bf5f-5a03a232fb08.png">

nssafir replied:
@deniederhut it seems to have shown up for me on the last build 
![image](https://user-images.githubusercontent.com/47368416/178109631-b95b3547-1c17-49ac-a0d0-d0385d4794b4.png)

deniederhut reviewed (changes requested):
Would you mind taking a second pass through your references section? The first three I looked at were all incomplete.

deniederhut commented:
@scoobies mark pending comment

stargaser commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21437/interspeech.2019-2848 is OK
- 10.1109/cvpr.2005.202 is OK
- 10.1109/cvpr.2017.713 is OK
- 10.1109/cvpr.2018.00552 is OK
- 10.1109/cvpr.2019.00482 is OK
- 10.1109/cvpr.2019.01108 is OK
- 10.1142/9789811223334_0120 is OK
- 10.1145/1015330.1015360 is OK
- 10.1016/j.patcog.2013.02.015 is OK
- 10.1007/978-3-030-93409-5_6 is OK
- 10.1371/journal.pone.0159161 is OK
- 10.25080/majora-342d178e-017 is OK
- 10.1016/j.ins.2021.05.016 is OK
- 10.5281/zenodo.3828935 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1109/cvpr.2015.729868 is INVALID
```

nssafir committed:
Add missing DOIs for new references

nssafir commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21437/interspeech.2019-2848 is OK
- 10.1109/cvpr.2005.202 is OK
- 10.1109/cvpr.2017.713 is OK
- 10.1109/cvpr.2018.00552 is OK
- 10.1109/cvpr.2019.00482 is OK
- 10.1109/cvpr.2019.01108 is OK
- 10.1142/9789811223334_0120 is OK
- 10.1145/1015330.1015360 is OK
- 10.1016/j.patcog.2013.02.015 is OK
- 10.1007/978-3-030-93409-5_6 is OK
- 10.1371/journal.pone.0159161 is OK
- 10.25080/majora-342d178e-017 is OK
- 10.1016/j.ins.2021.05.016 is OK
- 10.5281/zenodo.3828935 is OK
- 10.1007/978-3-030-18590-9_50 is OK
- 10.1109/ACCESS.2018.2835304 is OK
- 10.1016/j.neucom.2018.10.016 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1109/cvpr.2015.729868 is INVALID
```

nssafir committed:
Fix invalid FaceNet DOI

nssafir commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21437/interspeech.2019-2848 is OK
- 10.1109/cvpr.2015.7298682 is OK
- 10.1109/cvpr.2005.202 is OK
- 10.1109/cvpr.2017.713 is OK
- 10.1109/cvpr.2018.00552 is OK
- 10.1109/cvpr.2019.00482 is OK
- 10.1109/cvpr.2019.01108 is OK
- 10.1142/9789811223334_0120 is OK
- 10.1145/1015330.1015360 is OK
- 10.1016/j.patcog.2013.02.015 is OK
- 10.1007/978-3-030-93409-5_6 is OK
- 10.1371/journal.pone.0159161 is OK
- 10.25080/majora-342d178e-017 is OK
- 10.1016/j.ins.2021.05.016 is OK
- 10.5281/zenodo.3828935 is OK
- 10.1007/978-3-030-18590-9_50 is OK
- 10.1109/ACCESS.2018.2835304 is OK
- 10.1016/j.neucom.2018.10.016 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

nssafir commented:
@deniederhut thank you for bringing that to my attention, the issue with the DOIs should be resolved now.

nssafir committed:
Add doi's to all possible references

Also remove some unused references

nssafir commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.48550/ARXIV.1312.6114 is OK
- 10.21437/interspeech.2019-2848 is OK
- 10.48550/arXiv.1705.07120 is OK
- 10.1007/978-3-030-01267-0_42 is OK
- 10.48550/arXiv.2006.10102 is OK
- 10.1109/tpami.2013.50 is OK
- 10.1109/cvpr.2015.7298682 is OK
- 10.1109/cvpr.2005.202 is OK
- doi.org/10.1007/978-3-319-46478-7_31 is OK
- 10.48550/arXiv.2004.11362 is OK
- 10.1109/cvpr.2017.713 is OK
- 10.1109/cvpr.2018.00552 is OK
- 10.1109/cvpr.2019.00482 is OK
- 10.1109/cvpr.2019.01108 is OK
- 10.48550/arXiv.2105.05061 is OK
- 10.1142/9789811223334_0120 is OK
- 10.1145/1015330.1015360 is OK
- 10.1016/j.patcog.2013.02.015 is OK
- 10.48550/arXiv.2106.03609 is OK
- 10.1017/9781108955652.006 is OK
- 10.1007/978-3-030-93409-5_6 is OK
- 10.1371/journal.pone.0159161 is OK
- 10.48550/arXiv.2110.14795 is OK
- 10.25080/majora-342d178e-017 is OK
- 10.48550/arXiv.2008.10400 is OK
- 10.48550/arXiv.1907.08956 is OK
- 10.48550/arXiv.2008.00565 is OK
- 10.5281/zenodo.3828935 is OK
- 10.1007/978-3-030-18590-9_50 is OK
- 10.1109/ACCESS.2018.2835304 is OK
- 10.1016/j.neucom.2018.10.016 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1007/978-3-030-59430-5_13 is INVALID because of 'https://doi.org/' prefix
- 10.5555/2969442.2969628 is INVALID
- 10.5555/3157096.3157304 is INVALID
- 10.5555/1661445.1661640 is INVALID
- 10.5555/2969033.2969226 is INVALID
- 10.5555/1756006.1953024 is INVALID
- 10.5555/1953048.2078195 is INVALID
```

nssafir committed:
Remove one and fix one DOI

nssafir commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.48550/ARXIV.1312.6114 is OK
- 10.1007/978-3-030-59430-5_13 is OK
- 10.21437/interspeech.2019-2848 is OK
- 10.48550/arXiv.1705.07120 is OK
- 10.1007/978-3-030-01267-0_42 is OK
- 10.48550/arXiv.2006.10102 is OK
- 10.1109/tpami.2013.50 is OK
- 10.1109/cvpr.2015.7298682 is OK
- 10.1109/cvpr.2005.202 is OK
- doi.org/10.1007/978-3-319-46478-7_31 is OK
- 10.48550/arXiv.2004.11362 is OK
- 10.1109/cvpr.2017.713 is OK
- 10.1109/cvpr.2018.00552 is OK
- 10.1109/cvpr.2019.00482 is OK
- 10.1109/cvpr.2019.01108 is OK
- 10.48550/arXiv.2105.05061 is OK
- 10.1142/9789811223334_0120 is OK
- 10.1145/1015330.1015360 is OK
- 10.1016/j.patcog.2013.02.015 is OK
- 10.48550/arXiv.2106.03609 is OK
- 10.1017/9781108955652.006 is OK
- 10.1007/978-3-030-93409-5_6 is OK
- 10.1371/journal.pone.0159161 is OK
- 10.48550/arXiv.2110.14795 is OK
- 10.25080/majora-342d178e-017 is OK
- 10.48550/arXiv.2008.10400 is OK
- 10.48550/arXiv.1907.08956 is OK
- 10.48550/arXiv.2008.00565 is OK
- 10.5281/zenodo.3828935 is OK
- 10.1007/978-3-030-18590-9_50 is OK
- 10.1109/ACCESS.2018.2835304 is OK
- 10.1016/j.neucom.2018.10.016 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.5555/2969442.2969628 is INVALID
- 10.5555/3157096.3157304 is INVALID
- 10.5555/1661445.1661640 is INVALID
- 10.5555/1756006.1953024 is INVALID
- 10.5555/1953048.2078195 is INVALID
```

nssafir commented:
@deniederhut do you know why the DOIs with the `10.5555` prefix are showing up as invalid?

deniederhut commented:
Yeah that's weird. It looks like none of those have been registered with crossref. I would guess that they are internal IDs for ACM that just happen to look like DOIs?

<img width="625" alt="Screen Shot 2022-07-10 at 4 31 27 PM" src="https://user-images.githubusercontent.com/6868396/178162759-541c4105-68a8-45c6-b99e-86be73c95bfb.png">

It would be great if you could remove all the `10.5555` ones

nssafir committed:
Remove 10.5555 DOIs

nssafir commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.48550/ARXIV.1312.6114 is OK
- 10.1007/978-3-030-59430-5_13 is OK
- 10.21437/interspeech.2019-2848 is OK
- 10.48550/arXiv.1705.07120 is OK
- 10.1007/978-3-030-01267-0_42 is OK
- 10.48550/arXiv.2006.10102 is OK
- 10.1109/tpami.2013.50 is OK
- 10.1109/cvpr.2015.7298682 is OK
- 10.1109/cvpr.2005.202 is OK
- doi.org/10.1007/978-3-319-46478-7_31 is OK
- 10.48550/arXiv.2004.11362 is OK
- 10.1109/cvpr.2017.713 is OK
- 10.1109/cvpr.2018.00552 is OK
- 10.1109/cvpr.2019.00482 is OK
- 10.1109/cvpr.2019.01108 is OK
- 10.48550/arXiv.2105.05061 is OK
- 10.1142/9789811223334_0120 is OK
- 10.1145/1015330.1015360 is OK
- 10.1016/j.patcog.2013.02.015 is OK
- 10.48550/arXiv.2106.03609 is OK
- 10.1017/9781108955652.006 is OK
- 10.1007/978-3-030-93409-5_6 is OK
- 10.1371/journal.pone.0159161 is OK
- 10.48550/arXiv.2110.14795 is OK
- 10.25080/majora-342d178e-017 is OK
- 10.48550/arXiv.2008.10400 is OK
- 10.48550/arXiv.1907.08956 is OK
- 10.48550/arXiv.2008.00565 is OK
- 10.5281/zenodo.3828935 is OK
- 10.1007/978-3-030-18590-9_50 is OK
- 10.1109/ACCESS.2018.2835304 is OK
- 10.1016/j.neucom.2018.10.016 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

nssafir commented:
@deniederhut thanks for looking into that, I just removed the invalid DOIs

nssafir commented:
@deniederhut is all okay with this paper? I didn't see it listed in the proceedings so I just wanted to check if there were any further steps I should take.

stargaser commented:
I believe all the issues with the references have been addressed.

stargaser commented:
@scoobies mark ready

nssafir commented:
@stargaser does this PR need to be merged for the paper to be included in the proceedings?

deniederhut commented:
Sorry! I wasn't checking github during the conference. I'll take another look and merge if ready.

stargaser commented:
The paper will be merged and included in the final proceedings build by @deniederhut at the end of the month. As long as it is labeled as ready it is good for publication at that time.

deniederhut reviewed (approved):
This looks good -- thanks for the changes to the references!

nssafir commented:
I am still working on adding a discussion of the benchmarks, but everything
else has been addressed in some capacity, so the paper should be ready for
re review.

On Tue, Jun 28, 2022, 8:40 PM Dillon Niederhut ***@***.***>
wrote:

> @nssafir <https://github.com/nssafir> is this paper ready for re-review?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/695#issuecomment-1169416882>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALJMRYHOV5AORAY7Z4HHVFDVROLOJANCNFSM5XBTRNMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

deniederhut commented:
@nssafir did you reply to the wrong email? This conference ended several months ago.

nssafir commented:
My apologies, not sure why this comment sent!
